src/data_api/events.se

```diff
@@ -1,346 +1,420 @@
+def finalizeEvent(eventId):
+    if (!self.tryFinalizeEvent(eventId)):
+        ~invalid()
 
-def addReportersPaidSoFar(event):
-    self.controller.assertIsWhitelisted(msg.sender)
-    self.Events[event].reportersPaidSoFarForEvent += 1
+# will return 0 if event is not finalized by the time this method returns
+def tryFinalizeEvent(eventId):
+    # FIXME: handle finalizing forked events
+    # FIXME: when the event is finalized, we need to add its fees to the pool for the ending reporting window and we need to add the count of correct outcome tokens to the pool for the window.  These are necessary for fee collection
+    if (self.isEventFinalized(eventId)):
+        return(1)
+    if (block.timestamp < self.events[eventId].endTime):
+        return(0)
+    # if done with automated reporting window and there is no bond placed, finalize
+    nowIsAfterAutomatedReporterDisputeDue = block.timestamp > self.getAutomatedReportDisputeDueTimestamp(eventId)
+    eventHasAutomatedOutcome = self.events[eventId].automatedReporterOutcome != 0
+    eventHasNoAutomatedBondHolder = self.events[eventId].automatedReporterDisputeBondHolder == 0
+    if (nowIsAfterAutomatedReporterDisputeDue && eventHasAutomatedOutcome && eventHasNoAutomatedBondHolder):
+        self.events[eventId].finalOutcome = self.events[eventId].tentativeOutcome
+        return(1)
+    # migrate through forks in case that changes our scheduled reporting window
+    self.migrateEventThroughAllForks(eventId)
+    # if we aren't through the scheduled reporting window yet then the event can't be finalized
+    currentReportingWindow = self.controller.lookup('reporting').getCurrentReportingWindow()
+    scheduledReportingWindow = self.events[eventId].reportingWindow
+    nowIsBeforeOrDuringScheduledReportingWindow = currentReportingWindow <= scheduledReportingWindow
+    if (nowIsBeforeOrDuringScheduledReportingWindow):
+        return(0)
+    # disputes and following forks changes the reporting window, so if we make it to the end of a reporting window we know the event can be finalized
+    finalOutcome = self.events[eventId].tentativeOutcome
+    self.events[eventId].finalOutcome = finalOutcome
+    if (self.events[eventId].automatedReporterOutcome == finalOutcome && self.events[eventId].automatedReporterDisputeBondHolder != 0):
+        # TODO: pay bond holder 2x bond amount, pulling from incorrect reporting pool. NOTE: this needs to decrease available funds privateRedeemCorrectOutcomeTokens pulls from
+        return(1)
+    if (self.events[eventId].limitedReportersOutcome == finalOutcome && self.events[eventId].limitedReportersDisputeBondHolder != 0):
+        # TODO: pay bond holder 2x bond amount, pulling from incorrect reporting pool. NOTE: this needs to decrease available funds privateRedeemCorrectOutcomeTokens pulls from
+        return(1)
+    if (self.events[eventId].allReportersOutcome == finalOutcome && self.events[eventId].allReportersDisputeBondHolder != 0):
```

Comment:
If it is `!=` to final outcome we should pay the bond holder 2x the bond amount. If you're disputing an outcome and the final outcome was the same you should lose the bond, but if you're disputing and the final outcome was _different_ then your dispute was needed / worthwhile, so you get 2x the bond back.